### ISSUE TITLE
Dependency netresearch/jsonmapper makes it impossible to use Psalm in a software distributed as SaaS

### ISSUE THREAD

**Issue by kubawerlos** [vimeo/psalm#3303]
Psalm is [using](https://github.com/vimeo/psalm/blob/3.11.2/composer.json#L29) [netresearch/jsonmapper](https://github.com/cweiske/jsonmapper) which is distributes under OSL-3.0 [license](https://github.com/cweiske/jsonmapper/blob/master/LICENSE), which [claims](https://opensource.org/licenses/OSL-3.0):

> 5) External Deployment. The term "External Deployment" means the use, distribution, or communication of the Original Work or Derivative Works in any way such that the Original Work or Derivative Works may be used by anyone other than You, whether those works are distributed or communicated to those persons or made available as an application intended for use over a network. As an express condition for the grants of license hereunder, You must treat any External Deployment by You of the Original Work or a Derivative Work as a distribution under section 1(c).

and

> c) to distribute or communicate copies of the Original Work and Derivative Works to the public, with the proviso that copies of Original Work or Derivative Works that You distribute or communicate shall be licensed under this Open Software License;

@muglug would you consider replacing the library with some alternative? Or maybe @cweiske would consider changing the license?

P.S. I have found Magento is using the same [license](https://github.com/magento/magento2/blob/2.3.5/LICENSE.txt) and that surprised me.

**Comment by cweiske** [comment]
I chose OSL because it requires anyone to publicize changes made to the jsonmapper sources. It does not prevent you from using it in a SaaS product or anything; you only have to publish the changes you do to jsonmapper.

Github also shows you that: https://github.com/cweiske/jsonmapper/blob/master/LICENSE

>  Disclose source

**Comment by kubawerlos** [comment]
> It does not prevent you from using it in a SaaS product

You're saying this and I believe it is your intention, but that's not what the license describes:

 - the first part I quote (point 5.) about "External Deployment" is triggered when even using it in CI
 - that makes a "distribution under section 1(c)" to follow
 - 1(c) requires "to distribute or communicate copies of the Original Work and Derivative Works to the public"
 - and "Derivative Works" is the software that is using library with discussed license

In [this article](https://www.lawoftheledger.com/wp-content/uploads/sites/777/2018/08/Open-Source-and-SaaS.pdf) OSL 3.0 is compared to AGPL which was concern of [this](https://github.com/cweiske/jsonmapper/issues/6) issue.

**Comment by cweiske** [comment]
> and "Derivative Works" is the software that is using library with discussed license

No, this is not the case. To me, "Derivative works" means modified versions of jsonmapper. You only link to/use jsonmapper. I'm not sure what happens when you extend the JsonMapper class.

**Comment by kubawerlos** [comment]
The license describes - point 1(b):

> to translate, adapt, alter, transform, modify, or arrange the Original Work, thereby creating derivative works ("Derivative Works") based upon the Original Work;

or this is when I get it wrong and "Derivative Works" only applies to the library with discussed license, not a project build using it?

**Comment by cweiske** [comment]
This seems to be your problem :) Which of the verbs do you consider to be the one that extends it to the project?

**Comment by kubawerlos** [comment]
Kick me, I was thinking about the "adapt", but it turns out I need to (finally) learn Engish.

Today in the morning I was thinking the license is fine for using in ~SaaS~ any project.

Then after combining sections 1c and 5 I got convinced it is not ok, but immediately it started to bugging my why Magento is distributed with that license.

A few hours later - you know, I got older and smarter (and read your comments @cweiske) - I understood it only applies to the library distributed with the OSL 3.0 license, not the software built using it as a vendor library.

Anyway, thanks for the replies and sorry for the fuss.
